MaxEnt.py

```diff
@@ -1,181 +1,122 @@
-    for _ in range(iters):
-        # Generates a random array of indices of the shape (1, length |URs|).
-        training_order = np.random.choice(
-            a=range(len(violations)), size=len(violations), replace=False
-        )
-
-        dL_dw = 0
-
-        # Loops through each tableau in the dataset by random order.
-        for t in training_order:
-            # Skips the tableaux without a winning candidate.
-            if wIdx[t] == None:
-                continue
-
-            # Violation profile of the winning candidate.
-            winner_violations = np.array([violations[t][wIdx[t]]])
-
-            # Computes probability for each candidate given the current weights.
-            P = compute_probabilities(violations[t], w)
-
-            # Computes the gradient.
-            dL_dw += np.transpose(
-                winner_violations - (violations[t] * P).sum(axis=0, keepdims=True)
-            )
 
-        if reg:
-            # Adds the gradient of the sum difference prior.
-            total_w_fc = 0
-            total_w_mc = 0
 
-            for i, con in enumerate(constraints):
-                if con.startswith("*"):
-                    total_w_mc += w[i][0]
-                else:
-                    total_w_fc += w[i][0]
+class MaxEnt:
+    """A general-purpose MaxEnt class.
 
-            dP_dw = np.zeros(w.shape)
-            m_a = np.full(w.shape, mu)
-            s_a = np.full(w.shape, sigma)
-            total_w_fc_a = np.full(w.shape, total_w_fc)
-            total_w_mc_a = np.full(w.shape, total_w_mc)
+    Objects:
+        constraints (np.array[str]): Names of constraints.
+        weights (np.array[float]): Violation profiles for the tableaux.
 
-            for i, con in enumerate(constraints):
-                # For Markedness constraints:
-                if con.startswith("*"):
-                    dP_dw[i] = (total_w_mc_a[i] - total_w_fc_a[i] - m_a[i]) / s_a[i] ** 2
-                # For Faithfulness constraints:
-                else:
-                    dP_dw[i] = (total_w_fc_a[i] - total_w_mc_a[i] + m_a[i]) / s_a[i] ** 2
-
-            dL_dw += dP_dw
-
-        # Updates the weights.
-        w = w - eta * dL_dw
-
-        # Replaces negative weights with 0.
-        w[w < 0] = 0
-
-        # Store history.
-        history.append(w.squeeze())
-
-    return w, history
-
-
-def sorted_learned_weights(constraint, weight) -> dict[str, float]:
-    learned = {}
-    for cn, cw in zip(constraint, weight.squeeze()):
-        learned[cn] = np.round(cw, 3)
-
-    learned = sorted(learned.items(), key=lambda x: x[1], reverse=True)
+    Methods:
+        sorted_by_weights: Returns sorted (constraint, weight) tuples.
+        compute_probabilities: Returns the predicted probability distribution for any given violation profile.
+    """
 
-    for pair in learned:
-        print(f"{pair[0]:<20}: {pair[1]:<10}")
+    def __init__(self, constraints, weights):
+        self.const = constraints
```

Comment:
This is a super minor nitpick. but could we keep the variable naming consistent? i.e. cns / cws for constraint names / constraint names, or names / weights?